runtime-sdk/modules/evm/src/backend.rs

```diff
@@ -1,330 +1,526 @@
+        validator: V,
+    ) -> Result<subcall::SubcallResult, core::Error>;
+}
+
+impl<T: EVMBackendExt> EVMBackendExt for &T {
+    fn random_bytes(&self, num_bytes: u64, pers: &[u8]) -> Vec<u8> {
+        (*self).random_bytes(num_bytes, pers)
+    }
+
+    fn subcall<V: subcall::Validator + 'static>(
+        &self,
+        info: subcall::SubcallInfo,
+        validator: V,
+    ) -> Result<subcall::SubcallResult, core::Error> {
+        (*self).subcall(info, validator)
+    }
+}
+
+impl<'ctx, C: TxContext, Cfg: Config> EVMBackendExt for OasisBackend<'ctx, C, Cfg> {
+    fn random_bytes(&self, num_bytes: u64, pers: &[u8]) -> Vec<u8> {
+        // Refuse to generate more than 1 KiB in one go.
+        // EVM memory gas is checked only before and after calls, so we won't
+        // see the quadratic memory cost until after this call uses its time.
+        let num_bytes = num_bytes.min(RNG_MAX_BYTES) as usize;
+        let mut ctx = self.ctx.borrow_mut();
+        let mut rng = ctx.rng(pers).expect("unable to access RNG");
+        let mut rand_bytes = vec![0u8; num_bytes];
+        rand_core::RngCore::try_fill_bytes(&mut rng, &mut rand_bytes).expect("RNG is inoperable");
+        rand_bytes
+    }
+
+    fn subcall<V: subcall::Validator + 'static>(
+        &self,
+        info: subcall::SubcallInfo,
+        validator: V,
+    ) -> Result<subcall::SubcallResult, core::Error> {
+        let mut ctx = self.ctx.borrow_mut();
+
+        // Execute the subcall.
+        let mut result = subcall::call(&mut ctx, info, validator)?;
+        // Store state after subcall execution for substate handling.
+        self.pending_state
+            .borrow_mut()
+            .merge_from(mem::take(&mut result.state));
+
+        Ok(result)
+    }
+}
+
+pub struct OasisStackSubstate<'config> {
+    metadata: StackSubstateMetadata<'config>,
+    parent: Option<Box<OasisStackSubstate<'config>>>,
+    logs: Vec<Log>,
+    deletes: BTreeSet<H160>,
+    state: context::State,
+}
+
+impl<'config> OasisStackSubstate<'config> {
+    pub fn new(metadata: StackSubstateMetadata<'config>) -> Self {
+        Self {
+            metadata,
+            parent: None,
+            logs: Vec::new(),
+            deletes: BTreeSet::new(),
+            state: context::State::default(),
+        }
+    }
+
+    pub fn metadata(&self) -> &StackSubstateMetadata<'config> {
+        &self.metadata
+    }
+
+    pub fn metadata_mut(&mut self) -> &mut StackSubstateMetadata<'config> {
+        &mut self.metadata
+    }
+
+    pub fn enter(&mut self, gas_limit: u64, is_static: bool) {
+        let mut entering = Self {
+            metadata: self.metadata.spit_child(gas_limit, is_static),
+            parent: None,
+            logs: Vec::new(),
+            deletes: BTreeSet::new(),
+            state: context::State::default(),
+        };
+        mem::swap(&mut entering, self);
+
+        self.parent = Some(Box::new(entering));
+
+        CurrentStore::start_transaction();
```

Comment:
What do you mean? This is a critical piece of how this can work and the reason why we needed to implement `CurrentStore` in the first place. EVM stack executor manages its own state so support for raw storage transactions is mandatory.

Reply:
Was thinking of moving this higher up the calling stack. e.g. to `StackState`. This way the state would take care of the sub-states and the store.

Reply:
That's a good idea, updated.